opentelemetry-dynatrace/src/transform/metrics.rs

```diff
@@ -0,0 +1,1165 @@
+    }
+}
+
+impl From<MetricKey> for String {
+    /// Converts `MetricKey` instances into `String`.
+    fn from(key: MetricKey) -> Self {
+        key.0.into_owned()
+    }
+}
+
+impl fmt::Display for MetricKey {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut prev_char_underscore = false;
+        let mut escaped = self
+            .0
+            .as_ref()
+            .chars()
+            .filter_map(|c| {
+                if c == '.' || c == '-' || c == '_' || c.is_numeric() || c.is_ascii_alphabetic() {
+                    prev_char_underscore = false;
+                    Some(c)
+                } else if !prev_char_underscore {
+                    prev_char_underscore = true;
+                    Some('_')
+                } else {
+                    None
+                }
+            })
+            .peekable();
+
+        if escaped
+            .peek()
+            .map_or(false, |c| c == &'_' || c.is_ascii_alphabetic())
+        {
+            fmt.write_str(&escaped.take(250).collect::<String>())?;
+        } else {
+            fmt.write_str(&"_".chars().chain(escaped.take(249)).collect::<String>())?;
+        }
+
+        Ok(())
+    }
+}
+
+/// An immutable set of distinct metric dimensions.
+#[derive(Clone, Debug, Default)]
+pub struct DimensionSet {
+    dimensions: BTreeMap<Key, Value>,
+}
+
+impl DimensionSet {
+    /// The dimension set length.
+    pub fn len(&self) -> usize {
+        self.dimensions.len()
+    }
+
+    /// Check if the set of dimensions is empty.
+    pub fn is_empty(&self) -> bool {
+        self.dimensions.is_empty()
+    }
+
+    /// Iterate over the dimension key value pairs.
+    pub fn iter(&self) -> Iter<'_> {
+        self.into_iter()
+    }
+}
+
+impl fmt::Display for DimensionSet {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let val = self
+            .iter()
+            .enumerate()
+            .fold(String::new(), |mut acc, (idx, (key, value))| {
+                let offset = acc.len();
+                if idx > 0 {
+                    acc.push(',')
+                }
+
+                let mut prev_char_underscore = false;
+
+                let key = key
+                    .as_str()
+                    .chars()
+                    .filter_map(|c| {
+                        if c == '.'
+                            || c == '-'
+                            || c == '_'
+                            || c.is_numeric()
+                            || c.is_ascii_alphabetic()
+                        {
+                            prev_char_underscore = false;
+                            Some(c)
+                        } else if !prev_char_underscore {
+                            prev_char_underscore = true;
+                            Some('_')
+                        } else {
+                            None
+                        }
+                    })
+                    .take(100)
```

Comment:
Maybe worth commenting why we limit on 100 here. I understand it's because dimension only supports up to 50 pairs?

Reply:
I added comments to all occurrences of `take` to make it better understandable.